git/gogit/clone.go

```diff
@@ -363,36 +299,77 @@ func (g *Client) cloneSemVer(ctx context.Context, url, semverTag string, opts re
 	return buildCommitWithRef(cc, ref)
 }
 
-func (g *Client) cloneRefName(ctx context.Context, url string, refName string, cloneOpts repository.CloneOptions) (*git.Commit, error) {
+func (g *Client) cloneRefName(ctx context.Context, url string, refName string, opts repository.CloneOptions) (*git.Commit, error) {
 	if g.authOpts == nil {
 		return nil, fmt.Errorf("unable to checkout repo with an empty set of auth options")
 	}
+
 	authMethod, err := transportAuth(g.authOpts, g.useDefaultKnownHosts)
 	if err != nil {
 		return nil, fmt.Errorf("unable to construct auth method with options: %w", err)
 	}
-	head, err := getRemoteHEAD(ctx, url, plumbing.ReferenceName(refName), g.authOpts, authMethod)
-	if err != nil {
-		return nil, err
-	}
-	if head == "" {
-		return nil, fmt.Errorf("unable to resolve ref '%s' to a specific commit", refName)
-	}
 
-	hash := git.ExtractHashFromRevision(head)
 	// check if previous revision has changed before attempting to clone
-	if lastObserved := git.TransformRevision(cloneOpts.LastObservedCommit); lastObserved != "" {
-		if head != "" && head == lastObserved {
+	ref := plumbing.ReferenceName(refName)
+	if lastObserved := git.TransformRevision(opts.LastObservedCommit); lastObserved != "" {
+		head, err := getRemoteHEAD(ctx, url, ref, g.authOpts, authMethod)
+		if err != nil {
+			return nil, err
+		}
+		hash := git.ExtractHashFromRevision(head)
+		shortRef := fmt.Sprintf("%s@%s", refName, hash.Digest())
+		if head != "" && shortRef == lastObserved {
 			// Construct a non-concrete commit with the existing information.
 			c := &git.Commit{
-				Reference: refName,
 				Hash:      hash,
+				Reference: ref.String(),
 			}
 			return c, nil
 		}
 	}
 
-	return g.cloneCommit(ctx, url, hash.String(), cloneOpts)
+	var depth int
+	if opts.ShallowClone {
+		depth = 1
+	}
+	cloneOpts := &extgogit.CloneOptions{
+		URL:           url,
+		Auth:          authMethod,
+		RemoteName:    git.DefaultRemote,
+		ReferenceName: ref,
+		// TODO: go-git cloneRefSpec() doesn't handle SingleBranch == True with a ref like refs/pull/1/head because it calls .Short on the ref, and this cannot be converted to a short ref (no matching RefRevParseRules)
+		// I'm not quite sure how this works with SingleBranch == False either, because DefaultFetchRefSpec is "+refs/heads/*:refs/remotes/%s/*" which doesn't refs/pull/1/head either
+		// https://github.com/src-d/go-git/blob/v4.13.1/repository.go#L815-L835
+		// SingleBranch:      g.singleBranch,
```

Comment:
As mentioned in the PR message, go-git CloneOptions.SingleBranch does not support refs/pull/1/head because it builds the refpec to clone wrong.

I'm not entirely sure, but this may be a bug in go-git.

Ignoring singleBranch would be a regression for branches and other non-tag refs. Even with singleBranch=false go-git actually will detect tags automatically (it checks `ReferenceName.IsTag()` and only fetches one ref). I don't have a great fix to work around this. If it is a go-git bug it could be fixed upstream.